### PR TITLE
PT-4151: show "N notes couldn't be shown" in the comment list

### PR DESCRIPTION
**Pull request by rolfheij-sil**
## Update — 2026-07-13: rebased onto `main` + addressed @lyonsil's review

Rebased onto `main` (PT-4110 / #2524 merged). Then processed the `/code-review max` feedback:

**Implemented**
- **Comment-level drops are now counted (was thread-only).** `hiddenCount` now includes a comment dropped inside an otherwise-serializable thread, not just whole dropped threads — the realistically-reachable failure the ticket exists to surface. The `{ threads, hiddenCount }` envelope moved off the globally-registered `List<PlatformCommentThreadWrapper>` converter onto a dedicated `CommentThreadsResult` response type + converter, so the count spans both levels and a bare list of threads serialized elsewhere stays a plain array. `GetCommentThreads` still returns a queryable list; a thin wire wrapper produces the envelope. **Wire shape `{ threads, hiddenCount }` unchanged.**
- **Banner stays visible** on long lists (moved out of the scroll container) and is now a **polite** live region (`role="status"`).
- **Empty-state copy fixed**: when every matching thread is dropped, the misleading "No comments match filter" label is suppressed — the banner alone explains it.
- **Web-view result shape guarded** against a payload lacking `threads`.
- Tests: converter plural-path + `WriteIsolatedArray` nested-drop roll-up coverage.

**Deliberately not changed** (corner cases / larger work / soft): filter-scoped count, whole-response-error banner, Vitest for the React banner path, and the "1 comment" plural phrasing (matches the existing `"1 reply"` string). Per-item rationale is in the review threads.

---

# Code Review Summary

**Branch**: pt-4151-hidden-notes-indicator

**Base**: origin/pt-4110-isolate-note-serialization (stacked — this PR is #2531, based on PT-4110/#2524)

**Date**: 2026-07-10

**Review model**: Claude Opus 4.8

**Files changed**: 9

## Overview

PT-4151 adds a user-visible signal for the silent drops introduced by PT-4110. PT-4110 made `getCommentThreads` resilient — a thread that can't be serialized is dropped rather than failing the whole response — but the drop was invisible to the user. This change surfaces it: the C# serializer now returns a drop count, the `getCommentThreads` wire response carries it as `{ threads, hiddenCount }`, and the comment list renders a muted banner ("N comments couldn't be shown", en + es) when `hiddenCount > 0`, including above the empty-state when every thread was dropped. Scope: `hiddenCount` counts any note that can't be serialized — a whole dropped thread, or a single comment dropped inside an otherwise-healthy thread (see the 2026-07-13 update above; the original PR counted dropped *threads* only).

Four parallel analysis passes (api, style, compliance, ux) found no Critical issues, two Important, and two Minor. Three of the four were fixed during the review; the API one was dismissed as a deliberate, documented change.

## API Changes

- `extensions/src/legacy-comment-manager/src/types/legacy-comment-manager.d.ts`: Added exported type `LegacyCommentThreadsResult = { threads: LegacyCommentThread[]; hiddenCount: number }`.
- `extensions/src/legacy-comment-manager/src/types/legacy-comment-manager.d.ts`: Modified `LegacyCommentProjectInterfaceDataTypes.CommentThreads` data type — data shape `LegacyCommentThread[]` → `LegacyCommentThreadsResult`.
- `extensions/src/legacy-comment-manager/src/types/legacy-comment-manager.d.ts`: Modified `getCommentThreads()` return — `Promise<LegacyCommentThread[]>` → `Promise<LegacyCommentThreadsResult>`.
- `lib/papi-dts/papi.d.ts`: No diff (verified — `legacy-comment-manager` module types are not part of the papi bundle).
- `lib/platform-bible-react/`, `lib/platform-bible-utils/`: No changes.

### Critical — Must address before merge

None.

### Important — Should address before merge

- [x] **UX vocabulary: banner said "note(s)" while the whole panel says "comment(s)".** _(fixed during review: reworded the banner copy to "1 comment couldn't be shown" / "{count} comments couldn't be shown" in en + es, and renamed the keys `%comment_notes_hidden_*%` → `%comment_hidden_*%` so the identifier matches; the panel title, empty states, and type filter all use "comments", so the banner now matches.)_
- [ ] ~~**Wire-shape change to `CommentThreads` / `getCommentThreads` is technically breaking for the extension's public `declare module`.**~~ _(author: deliberate and it is the point of the ticket; verified no other in-repo consumers exist, the extension is `0.6.0-alpha.0`, and extension `tsc` is clean. Called out in the PR description so downstream extension authors aren't surprised. A non-breaking alternative — a second `HiddenCommentThreadCount` data type — was considered and rejected as heavier for no real gain at alpha.)_

[Author response: reworded the vocabulary as recommended; confirmed the wire-shape change is intentional and acceptable at alpha, and it is documented in the PR body.]

### Minor — Consider

- [x] **Duplicate `useMemo` guard in `comment-list.web-view.tsx`.** _(fixed during review: collapsed `safeCommentThreads` and `hiddenCount` into a single `useMemo<LegacyCommentThreadsResult>` that unwraps the result once — falling back to `DEFAULT_LEGACY_COMMENT_THREADS_RESULT` — then destructures `{ threads, hiddenCount }`.)_
- [x] **`WriteIsolatedArray` drop-count tests covered only 0 and 1 dropped, not >1.** _(fixed during review: added `WriteIsolatedArray_MultipleItemsThrow_CountsEachDrop`, which drops two elements and asserts `dropped == 2` with only the healthy elements written.)_

[Author response: both fixed during the interview.]

#### Shared Regions Modified

None. (No `#region shared with` markers in any changed file.)

#### Extension Config Changes

None. (All extension-side changes are source, tests, stories, and localization content — no build/config files touched.)

## Positive Observations

- The wire-shape change is threaded end-to-end in one PR — C# `WriteIsolatedArray` → `PlatformCommentThreadListConverter` → the `.d.ts` contract → the single in-repo TS consumer — with no stale references (verified via grep + clean extension `tsc`).
- `GetCommentThreads`'s C# method signature is deliberately unchanged (still `List<PlatformCommentThreadWrapper>`); only the JSON wire shape changes via the converter, keeping the C# contract and its list-based tests stable.
- Localization is complete and correctly wired: both keys added to `en` and `es` and registered in `COMMENT_LIST_PANEL_EXTRA_STRING_KEYS` (required for `useLocalizedStrings` to load them). No hardcoded UI text; the plural string reuses the established `formatReplacementString` utility rather than reinventing interpolation.
- The `Alert`/`AlertDescription`-only banner (no icon/title, default variant) matches the documented `ContentOnly` Alert pattern; sentence case and apostrophe style match the design guide.
- Storybook coverage matches the feature exactly — `OneNoteHidden` (singular), `SomeNotesHidden` (plural), `AllNotesHidden` (banner above the empty-state).
- C# tests updated in lockstep with the behavior change (drop-count + `{ threads, hiddenCount }` object-shape assertions, incl. the empty-list case).

## In-Review Quality Check

Changes were made during the review (the two Important/Minor fixes above). Afterward:
- `dotnet csharpier` — clean (no unrelated files reformatted).
- C# suite — **1409 passed / 0 failed / 6 skipped** (the new multi-drop test included).
- Extension `tsc --noEmit` — exit 0.
- `prettier --check` and `eslint` (`lint:scripts`) — clean on all changed files.

## Suggested Review Focus

- [ ] **Wire-shape change** — `CommentThreads` / `getCommentThreads` went from a bare `LegacyCommentThread[]` to `{ threads, hiddenCount }`. Deliberate and the only in-repo consumer is updated, but it is a breaking change to the extension's public module; confirm you're comfortable shipping it at alpha and that the PR-body call-out is sufficient for any downstream extension authors.

---

**Stacked on #2524** (`pt-4110-isolate-note-serialization`) — based on that branch and merges after it; GitHub retargets the base to `main` when #2524 merges. Jira: **PT-4151**.

AI-assisted (Claude Code) — designed and implemented subagent-driven (fresh implementer + review per task), then reviewed via `/review-paratext`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

- - -
This change is [Reviewable](https://reviewable.io/reviews/paranext/paranext-core/2531)

### CODE REVIEW

**Comment by lyonsil** [comment]
I don't have concerns with the wire-shape change mentioned in the suggested review focus.

I'm going to post `/code-review max` findings here because they do bring up some reasonable points, but I'm a little reluctant to drag this one out much. It is starting to get into corner cases of corner cases. So I don't think this is worth a lot of time, but if there are some simple adjustments that make sense to you based on the feedback, go ahead and make them.

File: c-sharp/JsonUtils/JsonConverterUtils.cs [hunk 1]
```diff
@@ -132,6 +135,7 @@ Func<T, string> describe
             writer.WriteRawValue(buffer.WrittenSpan, skipInputValidation: true);
         }
         writer.WriteEndArray();
+        return dropped;
```
Comment (lyonsil) [inline c-sharp/JsonUtils/JsonConverterUtils.cs:154]: **[correctness - high] Comment-level drops are silently uncounted.** This new `return dropped;` is consumed by `PlatformCommentThreadListConverter` (thread level), but the *other* caller - `PlatformCommentThreadConverter.cs:52`, which isolates individual comments within a thread - discards the return value. So if a comment's non-body getter throws (`PlatformCommentConverter` writes `value.User` / `value.VerseRefStr` / `value.Date` with no try/catch, unlike the body which degrades to a placeholder), that comment is dropped, the thread still serializes, `hiddenCount` stays 0, and no banner shows - the exact case PT-4151 exists to surface. A dropped comment also gets no 'could not be displayed' placeholder. Consider propagating the comment-level drop count into `hiddenCount`.

File: c-sharp-tests/JsonUtils/PlatformCommentThreadListConverterTests.cs [hunk 2]
```diff
@@ -31,29 +31,28 @@ private PlatformCommentThreadWrapper AddThread(Comment comment)
...
-    public void Serialize_ListWithOneUnserializableThread_DropsItAndKeepsTheRest()
+    public void Serialize_ListWithOneUnserializableThread_DropsItAndReportsHiddenCount()
     {
         PlatformCommentThreadWrapper goodThread = AddThread(CommentTestHelper.CreateBasicComment()); // 4217dff8
-        // An empty CommentThread (no comments) is a genuinely unserializable thread: reading its
-        // metadata (e.g. ModifiedDate, which indexes the last comment) throws mid-serialization. This
-        // exercises the real drop-and-discard path with no test-only production seam.
+        // An empty CommentThread (no comments) is genuinely unserializable: reading its metadata
+        // (e.g. ModifiedDate, which indexes the last comment) throws mid-serialization.
         var badThread = new PlatformCommentThreadWrapper(new CommentThread { ScrText = _scrText });
```
Comment (lyonsil) [inline c-sharp-tests/JsonUtils/PlatformCommentThreadListConverterTests.cs]: **[test-reachability - med] The only tested thread-drop trigger is filtered out before serialization.** `badThread` is an *empty* `CommentThread` (throws on `ModifiedDate`), but `ParatextProjectDataProvider.GetCommentThreads` drops threads without non-deleted comments (`.Where(t => t.HasNonDeletedComments)`, line 463) *before* the converter runs - so this trigger cannot reach the converter in production. No test exercises a thread that survives the upstream filter yet still fails thread-level serialization, so nothing proves `hiddenCount > 0` is reachable through the real pipeline. Combined with the comment-level gap, the counted (thread) path may rarely fire while real drops go uncounted.

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx [hunk 3]
```diff
@@ -230,6 +238,17 @@ export function CommentListPanel({
...
+            <AlertDescription>
+              {hiddenCount === 1
+                ? localizedStrings['%comment_hidden_single%']
+                : formatReplacementString(localizedStrings['%comment_hidden_multiple%'], {
+                    count: hiddenCount,
+                  })}
+            </AlertDescription>
+          </Alert>
+        )}
         {threads.length === 0 ? (
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: **[ux - med] Empty-state text misattributes the cause when all matching threads are dropped.** `noFiltersActive` (line 195) never consults `hiddenCount`. When a filter is active and every matching thread is dropped, this renders the banner *and* `%no_comments_match_filter%` ('No comments match filter') - which is false: the comments matched, they just could not be serialized. With no filter it shows 'No comments' beneath the banner. Consider suppressing/replacing the empty-state label when `hiddenCount > 0 && threads.length === 0`.

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx [hunk 4]
```diff
@@ -230,6 +238,17 @@ export function CommentListPanel({
 
       {/* Comments list */}
       <div className="tw:flex-1 tw:overflow-auto">
+        {hiddenCount > 0 && (
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: **[ux - med] Banner scrolls out of view on a long list.** The `Alert` is a non-sticky child inside the `tw:flex-1 tw:overflow-auto` scroll container (line 240), above `CommentList`. On a long thread list, scrolling down moves 'N comments couldn't be shown' off-screen, so a user who does not start at the top never sees it. Consider pinning it (sticky) or placing it outside the scroll region.

File: extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx [hunk 5]
```diff
@@ -199,7 +205,7 @@ global.webViewComponent = function CommentListWebView({
   // name resolves so the panel shows skeletons instead of that flash.
   const isAwaitingCurrentUserName = filters.assignment === 'assigned-to-me' && !currentUserName;
 
-  const [commentThreads, , isLoadingCommentThreads] = useProjectData(
+  const [commentThreadsResult, , isLoadingCommentThreads] = useProjectData(
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx:217]: **[ux - low/med] `hiddenCount` is filter-scoped, so narrowing a filter hides the warning.** The selector here is built from the active filters and applied in `GetCommentThreads` *before* serialization, so `hiddenCount` reflects only threads matching the current query. A corrupt note that raises the banner under 'All types' drops out of the count when the user switches to e.g. 'Conflicts' - the banner vanishes and the user can conclude nothing is wrong while the note stays hidden.

File: extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx [hunk 6]
```diff
@@ -213,13 +219,14 @@ global.webViewComponent = function CommentListWebView({
...
   );
 
-  const safeCommentThreads = useMemo<LegacyCommentThread[]>(() => {
-    if (!commentThreads || isPlatformError(commentThreads)) return [];
-    return commentThreads;
-  }, [commentThreads]);
+  const { threads: safeCommentThreads, hiddenCount } = useMemo<LegacyCommentThreadsResult>(() => {
+    if (!commentThreadsResult || isPlatformError(commentThreadsResult))
+      return DEFAULT_LEGACY_COMMENT_THREADS_RESULT;
+    return commentThreadsResult;
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx:244]: **[robustness - low] Unguarded shape assumption can crash the panel.** This memo returns `commentThreadsResult` when it is a truthy non-error value, then destructures `threads`. If a payload lacks `threads` (version skew - a newer extension against an older/overlaid C# provider still returning a bare array), `safeCommentThreads` is `undefined` and `CommentListPanel`'s `threads.length === 0` throws, crashing the web view. The pre-refactor code returned the array directly and could not fail this way. A shape check (or `?? DEFAULT`) would harden it.

File: extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx [hunk 7]
```diff
@@ -213,13 +219,14 @@ global.webViewComponent = function CommentListWebView({
...
-    DEFAULT_LEGACY_COMMENT_THREADS,
+    DEFAULT_LEGACY_COMMENT_THREADS_RESULT,
   );
 
-  const safeCommentThreads = useMemo<LegacyCommentThread[]>(() => {
-    if (!commentThreads || isPlatformError(commentThreads)) return [];
-    return commentThreads;
-  }, [commentThreads]);
+  const { threads: safeCommentThreads, hiddenCount } = useMemo<LegacyCommentThreadsResult>(() => {
+    if (!commentThreadsResult || isPlatformError(commentThreadsResult))
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx]: **[ux - low] Total-response `PlatformError` shows no banner.** On a whole-response failure this returns the default (`hiddenCount: 0`, `threads: []`), so the user sees 'No comments' with no banner and no error - the transparency goal fails exactly when everything is hidden. Likely pre-existing / out of scope, but worth a deliberate decision (e.g. an error state distinct from empty).

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx [hunk 8]
```diff
@@ -230,6 +238,17 @@ export function CommentListPanel({
...
+          <Alert className="tw:m-4 tw:w-auto">
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: **[a11y - low] Assertive banner re-announced on every reload.** `Alert` renders `role="alert"` (an assertive live region). Any filter change flips `isLoading` -> the panel renders the skeleton branch (unmounting the Alert) -> on reload with `hiddenCount > 0` it remounts and screen readers re-announce the message each time. Consider a persistent/polite presentation if this proves noisy.

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx [hunk 9]
```diff
@@ -230,6 +238,17 @@ export function CommentListPanel({
...
+            <AlertDescription>
+              {hiddenCount === 1
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: **[test-coverage - low] No Vitest coverage for the banner path or result unwrapping.** The `hiddenCount === 1 ? singular : plural` branch here and the web-view's `{ threads, hiddenCount }` unwrap (incl. the `isPlatformError`->default and shape-mismatch fallbacks) are validated only by Storybook stories, which `npm test` does not run or assert. Also the converter's `hiddenCount` is asserted only for 0 and 1 - the plural path (`>= 2`) the plural copy exists for is untested end-to-end. Flipping the boundary or reading the wrong field would pass CI.

File: c-sharp/JsonUtils/PlatformCommentThreadListConverter.cs [hunk 10]
```diff
@@ -23,11 +27,17 @@ public override void Write(
         Utf8JsonWriter writer,
         List<PlatformCommentThreadWrapper> value,
         JsonSerializerOptions options
-    ) =>
-        JsonConverterUtils.WriteIsolatedArray(
+    )
+    {
+        writer.WriteStartObject();
+        writer.WritePropertyName(THREADS);
+        int hiddenCount = JsonConverterUtils.WriteIsolatedArray(
```
Comment (lyonsil) [inline c-sharp/JsonUtils/PlatformCommentThreadListConverter.cs]: **[altitude - low/med] Diagnostic count is attached to a globally-registered collection converter, not the response/method.** The `{ threads, hiddenCount }` envelope is produced as a serialization side effect of `JsonConverter<List<PlatformCommentThreadWrapper>>`, registered globally in `SerializationOptions.cs`. Any future code that serializes a `List<PlatformCommentThreadWrapper>` (a nested field, a bulk export) will now silently emit an object where a bare array is expected, and the count cannot be logged/asserted server-side (`GetCommentThreads` still returns a bare list). A method-level response record would keep the diagnostic where it is produced and testable.

File: extensions/src/legacy-comment-manager/contributions/localizedStrings.json [hunk 11]
```diff
@@ -27,6 +27,8 @@
       "%comment_status_todo%": "Re-opened",
       "%comment_thread_multiple_replies%": "{count} replies",
       "%comment_thread_single_reply%": "1 reply",
+      "%comment_hidden_single%": "1 comment couldn't be shown",
```
Comment (lyonsil) [inline extensions/src/legacy-comment-manager/contributions/localizedStrings.json:33]: **[convention - low] Count copy uses the discouraged '1 comment' phrasing / binary plural.** Code-Style-Guide 'Localized Strings Guidelines' prefers 'Comments: 1' over '1 comment' and warns plural forms localize differently; the binary `=== 1 ? single : multiple` selection breaks CLDR languages with >2 plural forms once translated. Consistent with the existing `%comment_thread_single_reply%` strings, so soft - noting for awareness.

**Review by lyonsil — COMMENTED** [review]
Automated max-effort review of the PT-4151 changes (9 files, at head `3de1f832`). The array -> `{ threads, hiddenCount }` migration is cleanly propagated and there is no happy-path bug; the inline notes below are semantic / UX / test-coverage, ranked most-significant first.

Two earlier findings were already handled by the `address /review-paratext findings` commit (the two `useMemo`s were merged into one; a multi-drop `WriteIsolatedArray` test was added), so they are omitted.

Headline: the **comment-level vs thread-level drop-count gap** - `hiddenCount` counts only whole dropped threads, but the realistically-reachable failure (a comment dropped inside a surviving thread) is uncounted and unshown (see the `JsonConverterUtils.cs` and `PlatformCommentThreadListConverterTests.cs` comments).

File: c-sharp/JsonUtils/JsonConverterUtils.cs (hunk 1 above)
Comment (rolfheij-sil) [reply c-sharp/JsonUtils/JsonConverterUtils.cs:154]: 🤖 Claude: Fixed in `bb3fc9204c5`. `WriteThread` now **returns** its comment-drop count instead of discarding it, and `WriteIsolatedArray` takes an optional per-element writer whose nested drops roll up into the array total — so a comment dropped inside an otherwise-serializable thread now reaches `hiddenCount`. (Also moved the envelope off the global `List` converter; see the reply on `PlatformCommentThreadListConverter.cs`.)

File: c-sharp-tests/JsonUtils/PlatformCommentThreadListConverterTests.cs (hunk 2 above)
Comment (rolfheij-sil) [reply c-sharp-tests/JsonUtils/PlatformCommentThreadListConverterTests.cs]: 🤖 Claude: Agreed the empty-thread trigger can't reach the converter in production — it's a unit seam for exercising thread-level counting in isolation, which is still useful. The substantive gap you flagged (a comment dropped inside a surviving thread going uncounted) is the one now fixed in `bb3fc9204c5`, covered by new `WriteIsolatedArray` nested-drop roll-up tests plus a converter plural-path (`hiddenCount == 2`) test.

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx (hunk 3 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: 🤖 Claude: Fixed in `5f14e753bd3`. When `hiddenCount > 0 && threads.length === 0` the empty-state label is now suppressed; the banner alone explains the emptiness rather than falsely claiming "No comments match filter".

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx (hunk 4 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: 🤖 Claude: Fixed in `5f14e753bd3`. The banner moved out of the `tw:flex-1 tw:overflow-auto` scroll container so it stays pinned above the list instead of scrolling away.

File: extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx (hunk 5 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx:217]: 🤖 Claude: Acknowledged — not changed. `hiddenCount` is a byproduct of serializing the current filtered result set, so it necessarily reflects the active query. Surfacing drops across all filters would need a separate unfiltered serialization pass; deferring this as a corner case unless you'd like it pursued.

File: extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx (hunk 6 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx:244]: 🤖 Claude: Fixed in `69dd14fd0ab`. The unwrap memo now also checks `Array.isArray(...threads)` and falls back to the empty result, so a payload without a `threads` array can't leave `safeCommentThreads` undefined and crash the panel.

File: extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx (hunk 7 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/src/comment-list.web-view.tsx]: 🤖 Claude: Acknowledged — not changed (pre-existing / out of scope). A whole-response `PlatformError` still renders the empty state with no error. Distinguishing an error state from empty is a separate small feature; flagging as a deliberate deferral.

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx (hunk 8 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: 🤖 Claude: Fixed in `69dd14fd0ab`. The banner is now `role="status"` (polite) instead of the `Alert` default `role="alert"` (assertive), so filter reloads announce without interrupting or re-announcing loudly. The remount itself is inherent to the skeleton reload; the polite live region covers the noise.

File: extensions/src/legacy-comment-manager/src/comment-list.component.tsx (hunk 9 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/src/comment-list.component.tsx]: 🤖 Claude: Partially addressed in `bb3fc9204c5`. Added C# coverage for the plural (`>= 2`) path and the comment-drop roll-up (`WriteIsolatedArray` tests). Did **not** add Vitest/RTL for the React banner branch — Storybook covers it visually and the RTL harness is heavier than warranted here; happy to add it if you'd prefer.

File: c-sharp/JsonUtils/PlatformCommentThreadListConverter.cs (hunk 10 above)
Comment (rolfheij-sil) [reply c-sharp/JsonUtils/PlatformCommentThreadListConverter.cs]: 🤖 Claude: Fixed in `bb3fc9204c5`. Removed the globally-registered `JsonConverter<List<PlatformCommentThreadWrapper>>`; the `{ threads, hiddenCount }` envelope is now produced by a dedicated `CommentThreadsResultConverter` on a bespoke `CommentThreadsResult` response type, so an ordinary list of threads serialized elsewhere stays a plain array. `GetCommentThreads` still returns a queryable list (a thin wire wrapper builds the envelope) to avoid churning ~90 test call sites.

File: extensions/src/legacy-comment-manager/contributions/localizedStrings.json (hunk 11 above)
Comment (rolfheij-sil) [reply extensions/src/legacy-comment-manager/contributions/localizedStrings.json:33]: 🤖 Claude: Not changed. The binary "1 comment" / "{count} comments" form matches the sibling `%comment_thread_single_reply%` ("1 reply") string in this same file, and a "Comments: 1" form doesn't fit a full-sentence banner. Leaving it consistent with the existing strings, per your "soft" note.

File: c-sharp-tests/JsonUtils/CommentThreadsResultConverterTests.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    private PlatformCommentThreadWrapper AddThread(Comment comment)
+    {
+        _commentManager.AddComment(comment);
+        _commentManager.SaveUser(comment.User, false);
+        return new PlatformCommentThreadWrapper(_commentManager.FindThread(comment.Thread));
+    }
+
+    // An empty CommentThread (no comments) is genuinely unserializable: reading its metadata
+    // (e.g. ModifiedDate, which indexes the last comment) throws mid-serialization.
+    private PlatformCommentThreadWrapper UnserializableThread() =>
```
Comment (lyonsil) [inline c-sharp-tests/JsonUtils/CommentThreadsResultConverterTests.cs:35]: **[test-reachability] Suggest adding a thread whose corruption can actually occur on disk.**

Following up on my earlier `PlatformCommentThreadListConverterTests` thread rather than re-litigating it: this `UnserializableThread()` helper (an empty `CommentThread`) is now the trigger behind *both* `hiddenCount` assertions — the singular and the new plural one — and it's still the trigger you agreed can't reach the converter in production, because `GetCommentThreads` filters it out via `.Where(t => t.HasNonDeletedComments)` first. The new nested-drop coverage is proven on the synthetic `string[][]` / `SentinelStringConverter` harness in `JsonConverterUtilsTests`. So every drop test in the suite is still driven by data a real Notes file cannot contain, and nothing yet shows `hiddenCount > 0` is reachable end to end.

Also worth correcting something in my original review: I claimed the comment-level path was "the realistically-reachable failure." **That premise was wrong**, and it's my fault you built toward it. `PlatformCommentConverter` writes `User` / `VerseRefStr` / `Thread` / `Language` / `StartPosition` — which are plain **fields** on ParatextData's `Comment` and cannot throw. `Date` is `DateTime.ToString("o")`, `Id` is a `string.Format`, `ExtraHeadingInfo` is a null-safe split. Everything that *could* throw is already hardened by PT-4110 (enum conversions index-guarded, contents → placeholder, the five decode fields → `TryRender` → omitted, tags → `TryJoin`). I could not construct a realistic comment-level drop at all. The roll-up you added is correct and fine as defense-in-depth, but it isn't what will fire in production.

The reachable trigger is one level up: **a note whose `VerseRef` attribute doesn't parse.** `CommentThread.VerseRef` → `Comments[0].VerseRef` → `new VerseRef(str)` throws `VerseRefException`, and `PlatformCommentThreadConverter` writes `value.VerseRef.ToString()` unguarded — so the whole thread is dropped. Unlike the empty thread, this one *has* non-deleted comments, so it survives the upstream filter and actually reaches the converter.

Suggest adding this to `CommentThreadsResultConverterTests`. I ran it against `69dd14f` and it passes as written:

```csharp
[Test]
public void Serialize_ThreadWithCorruptVerseRef_DropsItAndReportsHiddenCount()
{
    // REALISTIC production trigger: a note whose VerseRef attribute in the Notes XML can't be
    // parsed (unknown book code here). CommentThread.VerseRef -> Comments[0].VerseRef ->
    // new VerseRef(str) throws VerseRefException, and PlatformCommentThreadConverter writes
    // VerseRef UNGUARDED -- so the whole thread is dropped. Unlike the empty-CommentThread
    // trigger, this thread HAS non-deleted comments, so it survives
    // ParatextProjectDataProvider's .Where(t => t.HasNonDeletedComments) filter and actually
    // reaches the converter in production.
    PlatformCommentThreadWrapper goodThread = AddThread(CommentTestHelper.CreateBasicComment()); // 4217dff8

    Comment corrupt = CommentTestHelper.CreateBasicComment();
    corrupt.Thread = "corrupt1";
    corrupt.VerseRefStr = "ZZZ 1:1"; // unknown book id -> VerseRef.Parse throws
    PlatformCommentThreadWrapper corruptThread = AddThread(corrupt);

    var json = Serialize(goodThread, corruptThread);

    using var doc = JsonDocument.Parse(json);
    JsonElement threads = doc.RootElement.GetProperty("threads");
    Assert.That(threads.GetArrayLength(), Is.EqualTo(1)); // corrupt thread dropped
    Assert.That(threads[0].GetProperty("id").GetString(), Is.EqualTo("4217dff8"));
    Assert.That(doc.RootElement.GetProperty("hiddenCount").GetInt32(), Is.EqualTo(1));
}
```

Output confirms the real path, not a seam:

```
WARNING: dropping comment thread corrupt1 from serialization; it could not be
serialized. Invalid reference : ZZZ 1:1
Passed! - Failed: 0, Passed: 6
```

This would be the first test in the file whose trigger is something a project could actually contain, and it's the only thing that would demonstrate the banner ever renders for a real user.

One thing it surfaced that I'll raise separately: the same corrupt note takes down the **entire** response once a scope filter is active — `FilterByScriptureRanges` (`ParatextProjectDataProvider.cs:1514`) reads `thread.VerseRef` unguarded inside the `Where` predicate, upstream of the serializer. Default scope is `UNFILTERED` so the banner works there, but under "current chapter"/"current book" the call throws, the web view falls back to the default empty result, and the user gets "No comments" with no banner and no error — healthy notes included. That's the deferred whole-response-error item, and it's the same data condition as this one.

**Review by lyonsil — COMMENTED** [review]
> @lyonsil partially reviewed 14 files and all commit messages, and resolved 11 discussions.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/2531)* status: all files reviewed, 1 unresolved discussion (waiting on rolfheij-sil).

File: c-sharp-tests/JsonUtils/CommentThreadsResultConverterTests.cs (hunk 12 above)
Comment (rolfheij-sil) [reply c-sharp-tests/JsonUtils/CommentThreadsResultConverterTests.cs:35]: 🤖 Claude: Following the thread — I turned your suggested trigger into a data-provider-level test (the real `GetCommentThreads` path, not just the converter), and it reframes this.

**The corrupt-VerseRef note throws earlier than the converter, and earlier than the scope filter.** With a healthy note also present, `GetCommentThreads` throws on its first line: `FindThreads` → `CommentManager.SortAndGroupThreads` → `Comment.CompareTo` → `Comment.VerseRef` → `new VerseRef("ZZZ 1:1")`. So the drop never reaches `PlatformCommentThreadConverter` (your test passes only because it hand-builds the wrapper and bypasses `FindThreads`), and it isn't scope-filter-specific — the sort runs on every call.

**But it can't happen in production at all.** `CommentManager.Load()` filters unparseable-VerseRef notes out at load:
```csharp
// Filter comments that have unparsable verse references
allComments.AddRange(commentList.Where(c => VerseRef.IsParseable(c.VerseRefStr)));
```
A note whose VerseRef doesn't parse never enters the CommentManager, so it can't reach `FindThreads`, the filter, or the serializer. My test only reproduced a throw by injecting via `AddComment`, which skips that filter.

So this trigger is in the same bucket as the comment-level one you retracted: not production-reachable. Between PT-4110's hardening, the `HasNonDeletedComments` filter, and this `IsParseable` load filter, I can't find any path that makes `hiddenCount > 0` for real data — which lands back on your original point that nothing demonstrates the banner rendering end to end.

Given that, the open question is bigger than a test: is the `hiddenCount` banner worth shipping as deliberate defense-in-depth (a cheap backstop if a future serializer field starts throwing, or if the load filter ever loosens), or should PT-4151 be reconsidered/descoped? What's your read?

**Scope, to be clear:** the `CommentThreadsResult` envelope, its converter, and the comment-level roll-up all exist only to carry `hiddenCount` — so if PT-4151 is descoped, the whole PR comes out together, not just the banner UI. (The bespoke type does also tidy up the #10 global-`List`-converter point, but that converter was itself PT-4151's first cut, so it isn't a reason to keep the PR.) The one piece that stands on its own — PT-4110's drop-don't-fail resilience — is already merged and unaffected either way.

**Comment by rolfheij-sil** [comment]
Closing this - YAGNE
